### DISCUSSION TITLE
SumoTraciInteractionApp

### DISCUSSION THREAD

**Discussion by jhvssimoes** [eclipse/mosaic#158] [Q&A] [answered]
Hello,

I'm new to Eclipse MOSAIC  and I'm having some trouble adapting the SumoTraciInteractionApp java application in the tutorials, to execute a different command. Specifically, I'm trying to do a Simulation Value Retrieval Distance Request (0xab 0x83). I want to determine the driving distance between two longitude and latitude coordinates. After reading the TraCI documentation and despite many hours trying to do this, it seems that i can't correctly create the byte array in the assembleTraciCommand method. Can anyone give me a hint on what is the byte array need to look like in the assembleTraciCommand method?

Thank you in advance.

**Comment by kschrab** [comment, marked as answer]
From reading the TraCI documentation, I would have set up the byte array like this:

```java
GeoPoint a = ...
GeoPoint b = ...

dos.writeByte(0xab); // simulation value retrieval
dos.writeByte(0x83); // distance Request
dos.writeInt(0); // simulation id (string with length 0)
dos.writeByte(0x0f); // type compound
dos.writeInt(3); // number of items of compound
dos.writeByte(0x01); // type: position
dos.writeDouble(a.toCartesian().getX());
dos.writeDouble(a.toCartesian().getY());
dos.writeByte(0x01); // type: position
dos.writeDouble(b.toCartesian().getX());
dos.writeDouble(b.toCartesian().getY());
dos.writeByte(1); //0 = air-distance, 1 = driving distance
```

**Comment by jhvssimoes** [comment]
Yes, that is very similar to most of my tries. Altough after setting up the array like that I most of the time end up with the following error on SUMO:

_Error: tcpip::Storage::readIsSafe: want to read 251658240 bytes from Storage, but only 36 remaining
Quitting (on error)._

I understand that we are sending 42 bytes in the array, but it seems that TraCI reads some of the first ones and then expects a large number of bytes and I don't know why is that.

Thank you for your reply.

**Reply by kschrab** [reply]
Ah, you need to add another string after the command varriable `0x83`. 

![image](https://user-images.githubusercontent.com/2386865/142642999-fbdb7afc-022c-40a7-a5e4-0f6fc3940ffb.png)

As this string is unused, you can pass an empty string, but you need to pass the length at least (which is `0` then):

```diff
  GeoPoint a = ...
  GeoPoint b = ...

  dos.writeByte(0xab); // simulation value retrieval
  dos.writeByte(0x83); // distance Request
+ dos.writeInt(0); // simulation id (string with length 0)
  dos.writeByte(0x0f); // type compound
  dos.writeInt(3); // number of items of compound
  dos.writeByte(0x01); // type: position
  dos.writeDouble(a.toCartesian().getX());
  dos.writeDouble(a.toCartesian().getY());
  dos.writeByte(0x01); // type: position
  dos.writeDouble(b.toCartesian().getX());
  dos.writeDouble(b.toCartesian().getY());
  dos.writeByte(1); //0 = air-distance, 1 = driving distance
```

indeed rather hard to find out :) I actually stumbled upon this when reading the python implementation in SUMO: https://github.com/eclipse/sumo/blob/270813b7fb8517ab288264297fe4c84e3d6ddb89/tools/traci/_simulation.py#L567

**Reply by jhvssimoes** [reply]
I was so close! I was filling that field with a string, but with length. Thank you very much!
